book/src/rfcs/ffadd.md

```diff
@@ -0,0 +1,228 @@
+
+$$a_2 + b_2 = q \cdot f_2 + r_2 - c_1$$
+
+
+Graphically, this is what is happening:
+
+```text
+bits  0..............87|88...........175|176...........263
+
+a  =  (-------a0-------|-------a1-------|-------a2-------)
++
+s = 1 | -1
+·
+b  =  (-------b0-------|-------b1-------|-------b2-------)
+                       >                >                >
+=                     c_0              c_1              c_2
+q  =  -1 |0 | 1
+·
+f  =  (-------f0-------|-------f1-------|-------f2-------)
++
+r  =  (-------r0-------|-------r1-------|-------r2-------)
+```
+
+
+Our witness computation is currently using the `BigUint` library, which takes care of all of the intermediate carry limbs itself so we do not have to address all casuistries ourselves (such as propagating the low carry flag to the high limb in case the middle limb is all zeros).
+
+### Upper bound check
+
+Last but not least, we should perform some range checks to make sure that the result $r$ is contained in $\mathbb{F}_f$. Ideally, we would like to reuse some gates that we already have. In particular, we can perform range checks for $0\leq X <2^{3\ell}=2^{3\cdot 88}$. But we want to check that $0\leq r<f$. The way we can tweak this gate to behave as we want, is the following. First, the above inequality is equivalent to saying that $-f \leq r - f < 0$. Then we add $2^{264}$ on both sides to obtain $2^{264} - f \leq r - f + 2^{264} < 2^{264}$. Thus, all there is left to check is that a given bound value is indeed computed correctly.
+
+This is very similar to a foreign field addition, but simpler: the overflow bit will always be $1$, the sign is positive $1$, and the right input is formed by the limbs $(0, 0, 2^{88})$ standing for the number $2^{264}$.  There could be intermediate limb carry bits $k_0$ and $k_1$ as above. Observe that, because the sum is meant to be $<2^{264}$, then the carry bit for the most significant limb should always be zero $k_2 = 0$, so we do not use it. Happily, we can apply the addition gate again to perform the addition limb-wise, by selecting the following parameters:
+
+$$
+\begin{aligned}
+a_0 &=& r_{min_{0}}, \\ a_1 &=& r_{min_{1}}, \\ a_2 &=& r_{min_{2}} , \\
+b_0 &=& 0, \\ b_1 &=& 0, \\ b_2 &=& 2^{88} , \\
+s &=& 1 \\
+o &=& 1
+\end{aligned}
+$$
+
+Calling $u$ the upper bound term, the equation $r + 2^{264} - f $ can be expressed as $r + 2^{264} = 1 * f + u$. Finally, we perform a range check on the sum $u$, and we would know that $r < f$.
+
+
+```text
+bits  0..............87|88...........175|176...........263
+
+r  =  (-------r0-------|-------r1-------|-------r2-------)
++
+g  =  (-------g0-------|-------g1-------|-------g2-------)
+                       >                >
+=                     k_0              k_1
+f
++
+u  =  (-------u0-------|-------u1-------|-------u2-------)
+```
+
+Following the steps above, and representing this equation in limb form, we have:
+
+$$
+\begin{aligned}
+u_0 &= r_0 + 0 - k_0 \cdot 2^{88} \\
```

Comment:
Actually no, that is the cool point of reusing the gate this way. Let me give some intuition. The operation we are doing here is `r + 2^264 - f`. This will **always** produce overflow, right? But not only it will be larger than `f`, but it will be around (but less) `2^264`. What I mean is that the upper bound will not have the shape that outputs normally have (in the sense that `r` is meant to always be `<f`). We could say that the modulus term is "embedded" into the upper bound. 

Let's look at it from a different angle. What you're suggesting could make sense IF the right input of this final FFAdd gate was `2^264 - f`, but the thing is it is not! We are just considering right to be `2^264`.  Then one could say, ok, but where is the `f` term going? Mmm....

If we just do `r_min + 2^264`, which is what we are indeed doing, this will for sure produce an overflow, right? Ok, this means the FFAdd gate will be able to express `r_min + 2^264 = 1 * f + something`. Call this something `u`. Voilà! You just have `u = r_min + 2^264 - f` for free! Which is what you wanted in the first place ;)